### PR TITLE
iki_tukr

### PR DESCRIPTION
**Pull request by ryusei-iki**
minimum 
- [x] 関係データについて理解し、関係データとは何か説明できる。
- [x] TUKRのタスク設定について理解し、数式やアルゴリズムの各ステップを説明できる。
- [x] data_scrach.pyの穴埋めを行い関係データの人工データを作成する。
- [x] UKRの実装ファイルを改良してTUKRを実装する。

standard
- [ ] 欠損データを実装する
- [ ] 手動微分で実装する。
- [ ] 描画方法を変更する
- [ ] クロステストの実装
- [ ] 実データでの実装。(動物データやNIPSデータなど)

extra
- [ ] 手動微分を数式的に導くことができる。

### CODE REVIEW

File: Lecture_TUKR/utils/show_images_tukr.py [hunk 1]
```diff
@@ -4,11 +4,13 @@
 import jax,jaxlib
 import jax.numpy as jnp
 import numpy as np
-
+import sys,os
+moto=os.getcwd()[:-5]+'iki/'
+sys.path.append(moto)
```
Comment (TetraMiyazaki) [inline Lecture_TUKR/utils/show_images_tukr.py:9]: このへんこまってるなら pathっていう3.8系くらいから追加された標準ライブラリ結構使えるかもです．

**Comment by TetraMiyazaki** [comment]
Pull Requestの名前も書いといてください

**Comment by nahoando** [comment]
ミニマム、スタンダード、エクストラの設定をまずしましょう。何か結果が出た場合や作業が滞った場合などは結果をシェアして先輩や同級生のアドバイスもらえる環境づくり心がけましょう。実際こちらもコード見ただけではわからないところも多いです。add, commit, pushを忘れずに！

**Comment by ryusei-iki** [comment]
minimum 
- [ ] 関係データについて理解し、関係データとは何か説明できる。
- [ ] TUKRのタスク設定について理解し、数式やアルゴリズムの各ステップを説明できる。
- [ ] data_scrach.pyの穴埋めを行い関係データの人工データを作成する。
- [ ] UKRの実装ファイルを改良してTUKRを実装する。

standard
- [ ] 欠損データを実装する
- [ ] 手動微分で実装する。
- [ ] 描画方法を変更する
- [ ] クロステストの実装
- [ ] 実データでの実装。(動物データやNIPSデータなど)

extra
- [ ] 手動微分を数式的に導くことができる。

**Comment by ryusei-iki** [comment]
動かして描画するところまでは出来ましたが，上手く学習が行かなかったです．

Z1のカーネルとZ2のカーネルをかけるところが怪しいです．


学習時の描画

https://user-images.githubusercontent.com/103087980/169459880-880c7efa-4c55-4cc0-85d0-6e0c53da8615.mp4



ワイヤーフレームの描画

https://user-images.githubusercontent.com/103087980/169459790-ee0f2a75-6ddb-44af-86ab-3c4d5c46d555.mp4

潜在空間Z1の描画

https://user-images.githubusercontent.com/103087980/169459826-efc0d62c-a194-4a4a-8b94-7c85537656b1.mp4


潜在空間Z2 の描画

https://user-images.githubusercontent.com/103087980/169459854-dbdc289c-e680-4907-b1f9-c31b166b1abb.mp4

すべての損失
loss

二乗誤差の損失
loss_mse

正則化項の損失
loss_L2

File: Lecture_TUKR/iki/tukr_gauss/tukr.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+                    k = jnp.exp(-1 / (2 * self.sigma ** 2) * d)
+                    # print(d.shape)
+            return k
+
+        def karnel_jnp(target1,target2,type=1):
+            k1= kernel_f(target1,type,1)
+            k2=kernel_f(target2,type,2)
+            # print(target1.shape)
+            # print(k1.shape,k2.shape)
+            k=jnp.einsum('ij,jk->ik',k1,k2)
```
Comment (hirowatari-s) [inline Lecture_TUKR/iki/tukr_gauss/tukr.py:110]: 110と111行目を考え直してみてください．
まず，110行目でkを2階のテンソルに丸めていますが，この次の式でxijと掛けなきゃなのにこれだとij成分が消えちゃってますね〜

**Comment by TetraMiyazaki** [comment]
```
すべての損失 loss
二乗誤差の損失 loss_mse
正則化項の損失
```

このへんみてるのはナイスですね〜
書いてないけど，正則化項の微分しか効いてないって推測して写像の方に問題があるって思っているのは正しい．ですね．

**Comment by hirowatari-s** [comment]
あと実装とは関係ないですが，Assignees（このプルリクの担当者）はイキくんですね！

File: Lecture_TUKR/iki/tukr_gauss/tukr.py (hunk 2 above)
Comment (TetraMiyazaki) [reply Lecture_TUKR/iki/tukr_gauss/tukr.py:110]: これは，取り組み方ですが，数式にでてくる項を変数だとどのようなshapeになるか書きながらやってみるといいかもです．

File: Lecture_TUKR/iki/tukr_gauss/tukr.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+        # m_x = m_x.reshape(-1)
+        # m_y = m_y.reshape(-1)
+        return z1,z2
+
+    def fit(self,):
+        #テスト時に入力がznのときのyを求める関数
+        def kernel_f(target,type=1,jyun=-1):
+            if(type==1):
+                d = jnp.sum((target[:, None, :] - target[None, :, ]) ** 2, axis=2)
+                k = jnp.exp(-1 / (2 * self.sigma ** 2) * d)
```
Comment (TetraMiyazaki) [inline Lecture_TUKR/iki/tukr_gauss/tukr.py:90]: これ，self書くのがめんどくさいからこういう書き方してるんかな？（はじめてみた）
外部から参照が入らない場合はこれでもいいかもね．（違和感すごいけど）

**Comment by ryusei-iki** [comment]
レコード型のTUKRを取り組んでいます．

計算式は間違っていないと思う．
ワイヤーフレームは上手く行っていない．
潜在空間を二次元にしているが，一次元的な動きしかしていない．

ワイヤーフレームが上手く行っていない理由は，潜在空間上の点が線上になっているため，二次元のメッシュグリッドを潜在空間として与えても教師データにない点が多いためだと思う．

次は，潜在空間を1次元にして試してみる．

https://user-images.githubusercontent.com/103087980/170452657-484404ec-2716-41b6-83a2-8ddc9345c5a8.mp4


https://user-images.githubusercontent.com/103087980/170452720-d386590b-cb62-43f1-b343-0f406b11d1c2.mp4


https://user-images.githubusercontent.com/103087980/170452748-83030b10-1df6-4dfd-a849-0ad488c6b93e.mp4


https://user-images.githubusercontent.com/103087980/170452759-92bb7797-331f-4a25-b5d8-3c0961fd9453.mp4

**Comment by TetraMiyazaki** [comment]
> 計算式は間違っていないと思う．ワイヤーフレームは上手く行っていない．

学習ができているかの確認の段階はまずはxの推定値を描画してみるといいと思います．単純に与えられたデータに近くにきているかでまず判断しやすいので．

> 潜在空間を二次元にしているが，一次元的な動きしかしていない．
この潜在空間の話は直積潜在空間$\mathcal{U} \times \mathcal{V}$の空間の話でしょうか．それとも個別の潜在空間$\mathcal{U}$の話でしょうか

**Comment by ryusei-iki** [comment]
>> 計算式は間違っていないと思う．ワイヤーフレームは上手く行っていない．

>学習ができているかの確認の段階はまずはデータの推定値$\hat{x}_n = f(z_n)$を描画してみるといいと思います．単純に与えられたデータに近くにきているかでまず判断しやすいので．

y_zn.mp4がデータの推定値とデータとの比較の動画だと思います．これを見る限り，学習は出来ている！？

>> 潜在空間を二次元にしているが，一次元的な動きしかしていない．

>この潜在空間の話は直積潜在空間$\mathcal{U} \times \mathcal{V}$の空間の話でしょうか．それとも個別の潜在空間$\mathcal{U}$の話でしょうか

個別の潜在空間の話です．

**Comment by TetraMiyazaki** [comment]
> y_zn.mp4がデータの推定値とデータとの比較の動画だと思います．これを見る限り，学習は出来ている！？

あ，本当ですね．これをみると，$x_{ij} \simeq f(u_i, u_j)$という学習はできてますよね．

**Comment by TetraMiyazaki** [comment]
> 個別の潜在空間の話です．

なるほど，$\mathcal{U} \times \mathcal{V}$の直積空間の次元数はそれぞれ２と２だったら４次元になりますね．つまり，４次元多様体でfittingするような感じになります．一方でそれぞれの次元数が１と１だったら２次元になります．

**Comment by ryusei-iki** [comment]
u，vの潜在空間を両方とも一次元にしてみました．
そしたら，線にしかならなかったです．

潜在空間は，uの最小値から最大値のlinspace 100，vの最小値から最大値のlinspace 100としている．

https://user-images.githubusercontent.com/103087980/170495023-7e7e4c09-eadf-4ff7-bbc7-f16b9d1013b7.mp4



計算式が間違っているかと思い，学習に用いた潜在空間を使ってみたら，同じところに出来たので，計算式は間違っていないと思う．


https://user-images.githubusercontent.com/103087980/170493858-503fc12b-13b0-45c8-b9f0-995798131aca.mp4



https://user-images.githubusercontent.com/103087980/170493887-4cbefb35-2994-4873-8d4f-d43787266046.mp4

最初のやりかたでは潜在空間が関係データっぽくなかったので，学習時の潜在空間ををソートして（[[9,2][2,2][9,1][2,1]]とあったら，[[2,1][2,2][9,1][9,2]]）としたら，また線になった．欠損のない関係データにしているから，ソートをしても組み合わせは変わらずに順番が変わるだけだから，少なくとも，y_znのように点が散らばると思ったが，そうではなかった．

https://user-images.githubusercontent.com/103087980/170494827-05c9ef21-dfd0-4cd0-97af-81ed143f93e5.mp4


https://user-images.githubusercontent.com/103087980/170494862-8bb34567-50a5-4359-a21c-55f05b1bfbd7.mp4

**Comment by TetraMiyazaki** [comment]
@ryusei-iki 
今学習した潜在変数$\hat{z}$を入力とした$\hat{f}(\hat{z})$場合は問題がなく，新しく作った$\zeta$の$\hat{f}(\zeta)$写像がうまくいってないという状態ですよね．
学習の時に作った真の潜在変数を入力してみてそれでも$\hat{f}(\hat{z_{true}})が$うまくいかない場合は，学習がうまくいってない or 描画が間違っている．
うまくいく場合は，新しく作ったzetaの作り方が間違っている or なにか起きてる みたいな感じで推測できると思うので，
真の潜在変数を入力してみると色々わかるかもです．

**Comment by ryusei-iki** [comment]
>最初のやりかたでは潜在空間が関係データっぽくなかったので，学習時の潜在空間ををソートして（[[9,2][2,2][9,1][2,1]]とあったら，[[2,1][2,2][9,1][9,2]]）としたら，また線になった．欠損のない関係データにしているから，ソートをしても組み合わせは変わらずに順番が変わるだけだから，少なくとも，y_znのように点が散らばると思ったが，そうではなかった．

学習時の潜在変数をソートしたらワイヤーフレームは表示することはできた．
ただソートを間違っていただけだった．なぜ線になってたかというと，片方の潜在変数が一つの値を繰り返してたから．だから，zetaで線になるのも同じ理由かも知れないから確認する．



https://user-images.githubusercontent.com/103087980/170606285-24d9532a-a52a-463e-8c18-3986728297c9.mp4

https://user-images.githubusercontent.com/103087980/170606459-77e89658-36d1-4cfd-aa32-2ad91ca44c8c.mp4


https://user-images.githubusercontent.com/103087980/170606471-e7128bd9-20f1-4f3f-a7a5-27500083c420.mp4

**Comment by ryusei-iki** [comment]
>ただソートを間違っていただけだった．なぜ線になってたかというと，片方の潜在変数が一つの値を繰り返してたから．だから，zetaで線になるのも同じ理由かも知れないから確認する．

zetaを確認したら，片方の潜在変数が一つの値を繰り返してはいなかった．他に線になる理由を見つけた．uとvを両方とも単調増加させてペアにしてたのが線になる理由だと思う．zetaを100にするとしたら，今まではnp.arange(100)にしていた．これをnp.arange(10)を作って，その組み合わせで100個のペアを作った．そしたら出来た．


https://user-images.githubusercontent.com/103087980/170608321-974154fc-0388-4515-b57e-bb7eaee4be38.mp4



https://user-images.githubusercontent.com/103087980/170608337-3e8a66bc-4a49-4ddb-8a9b-87d68a78438d.mp4




https://user-images.githubusercontent.com/103087980/170608353-9db20040-546e-47a1-bb4a-2f84bc1d429a.mp4

**Comment by TetraMiyazaki** [comment]
> uとvを両方とも単調増加させてペアにしてたのが線になる理由だと思う．zetaを100にするとしたら，今まではnp.arange(100)にしていた．これをnp.arange(10)を作って，その組み合わせで100個のペアを作った．そしたら出来た．

なるほど！！！！ナイファイっす！！！ほんとに．

**Comment by TetraMiyazaki** [comment]
徳永くんにもいったんですが，いきくんもvisualizerのprojectに関わってるので，
ミニマムが終わり次第すぐに，スタンダードサクセスの描画方法を変更する．のCCPの実装を優先して取り掛かりましょう．

**Comment by ryusei-iki** [comment]
TUKRのtensorでの実装が出来ました．

https://user-images.githubusercontent.com/103087980/171762876-d4d92bcf-8853-4ef0-95f9-a650c1293d6e.mp4


https://user-images.githubusercontent.com/103087980/171762891-f906a62d-ad84-437d-982c-bbe6ddb16cff.mp4


https://user-images.githubusercontent.com/103087980/171762898-a2027773-fe77-4c51-b20b-cb7810c21bd9.mp4

**Comment by hirowatari-s** [comment]
さすが！！！いい感じですね！
潜在空間のプロットは範囲を-1から1に固定するのがおすすめです！
あと誤差関数をプロットしてみて減少しているか確認してみるといいかもです．

**Comment by ryusei-iki** [comment]
潜在空間を-1から1にすると，見やすくなりました．

https://user-images.githubusercontent.com/103087980/171765062-fd593d8c-cb16-4ce8-bc97-ffd54af04baa.mp4

https://user-images.githubusercontent.com/103087980/171765082-583ce423-24ff-4b54-bd40-54574513e8f2.mp4

損失はこんな感じで減っていました．
20epochぐらいの動きが良く分かっていないです．

loss
loss_mse
loss_L2

**Comment by hirowatari-s** [comment]
損失3枚プロットしてくれてるけど，どう違うやつかな？

**Comment by ryusei-iki** [comment]
>あ，全てと二乗誤差と正則化誤差かな？

そうです．
>３０エポックぐらいまでをゆっくり観測空間と潜在空間プロットしてみたらなんか見えるかも？

やってみます．

**Comment by TetraMiyazaki** [comment]
> 20epochぐらいの動きが良く分かっていないです．

広渡さんもいってますが，ここに着目するのはかなりいいと思うので，観察してみるとよいです

**Review by TetraMiyazaki — CHANGES_REQUESTED** [review]
観測空間の色付けとここの潜在空間のの色付けの対応がわからないです．
UKRのときはちゃんとデータの位相がとれているかを色付けで確認しましたよね．関係データのほうもちゃんと位相がとれているかを色付けで示すようにしてください．
```suggestion
        self.ZN1_calc=np.zeros((self.X1_num*self.X2_num,self.L))
```
